diff --git a/System/FilePath.hs b/System/FilePath.hs
index 331ae81..0c75d25 100644
--- a/System/FilePath.hs
+++ b/System/FilePath.hs
@@ -20,7 +20,10 @@ functions.
 -}
 
 
-#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+#if defined(ghcjs_HOST_OS)
+module System.FilePath (module System.FilePath.Current) where
+import System.FilePath.Current
+#elif defined(mingw32_HOST_OS) || defined(__MINGW32__)
 module System.FilePath(module System.FilePath.Windows) where
 import System.FilePath.Windows
 #else
diff --git a/System/FilePath/Current.hs b/System/FilePath/Current.hs
new file mode 100644
index 0000000..0355326
--- /dev/null
+++ b/System/FilePath/Current.hs
@@ -0,0 +1,4 @@
+{-# LANGUAGE CPP #-}
+#define MODULE_NAME     Current
+#define IS_CURRENT      1
+#include "Internal.hs"
diff --git a/System/FilePath/Internal.hs b/System/FilePath/Internal.hs
index 051e44c..6d7d61f 100644
--- a/System/FilePath/Internal.hs
+++ b/System/FilePath/Internal.hs
@@ -1,4 +1,6 @@
-#if __GLASGOW_HASKELL__ >= 704
+#ifdef IS_CURRENT
+{-# LANGUAGE ForeignFunctionInterface, JavaScriptFFI, Trustworthy #-}
+#elif __GLASGOW_HASKELL__ >= 704
 {-# LANGUAGE Safe #-}
 #endif
 {-# LANGUAGE PatternGuards #-}
@@ -63,6 +65,8 @@
 -- [1] <http://msdn.microsoft.com/en-us/library/windows/desktop/aa365247.aspx Naming Files, Paths and Namespaces> (Microsoft MSDN)
 module System.FilePath.MODULE_NAME
     (
+    isWindows, isPosix,
+    
     -- * Separator predicates
     FilePath,
     pathSeparator, pathSeparators, isPathSeparator,
@@ -124,8 +128,11 @@ isPosix = not isWindows
 
 -- | Is the operating system Windows like
 isWindows :: Bool
+#ifdef IS_CURRENT
+foreign import javascript unsafe "h$filepath_isWindows()" isWindows :: Bool
+#else
 isWindows = IS_WINDOWS
-
+#endif
 
 ---------------------------------------------------------------------
 -- The basic functions
diff --git a/filepath.cabal b/filepath.cabal
index 0841a0d..9c1295a 100644
--- a/filepath.cabal
+++ b/filepath.cabal
@@ -40,6 +40,9 @@ library
         System.FilePath
         System.FilePath.Posix
         System.FilePath.Windows
+    if impl(GHCJS>=0.1)
+        Exposed-modules:
+            System.FilePath.Current
 
     build-depends:
         base >= 4 && < 4.9
