diff --git a/Data/Hashable/Class.hs b/Data/Hashable/Class.hs
index f290f6f..367c1e5 100644
--- a/Data/Hashable/Class.hs
+++ b/Data/Hashable/Class.hs
@@ -89,7 +89,7 @@ import qualified Data.ByteString.Lazy.Internal as BL  -- foldlChunks
 import qualified Data.ByteString.Short.Internal as BSI
 #endif
 
-#ifdef VERSION_integer_gmp
+#if defined(VERSION_integer_gmp) && !defined(ghcjs_HOST_OS)
 
 # if MIN_VERSION_integer_gmp(1,0,0)
 #  define MIN_VERSION_integer_gmp_1_0_0
@@ -258,7 +258,7 @@ instance Hashable Char where
     hash = fromEnum
     hashWithSalt = defaultHashWithSalt
 
-#if defined(MIN_VERSION_integer_gmp_1_0_0)
+#if defined(MIN_VERSION_integer_gmp_1_0_0) && !defined(ghcjs_HOST_OS)
 instance Hashable BigNat where
     hashWithSalt salt (BN# ba) = hashByteArrayWithSalt ba 0 numBytes salt
                                  `hashWithSalt` size
@@ -269,7 +269,7 @@ instance Hashable BigNat where
 
 #if MIN_VERSION_base(4,8,0)
 instance Hashable Natural where
-# if defined(MIN_VERSION_integer_gmp_1_0_0)
+# if defined(MIN_VERSION_integer_gmp_1_0_0) && !defined(ghcjs_HOST_OS)
     hash (NatS# n)   = hash (W# n)
     hash (NatJ# bn)  = hash bn
 
@@ -283,7 +283,7 @@ instance Hashable Natural where
 #endif
 
 instance Hashable Integer where
-#if defined(VERSION_integer_gmp)
+#if defined(VERSION_integer_gmp) && !defined(ghcjs_HOST_OS)
 # if defined(MIN_VERSION_integer_gmp_1_0_0)
     hash (S# n)   = (I# n)
     hash (Jp# bn) = hash bn
