diff --git a/Data/Hashable/Class.hs b/Data/Hashable/Class.hs
index da4ad90..2145974 100644
--- a/Data/Hashable/Class.hs
+++ b/Data/Hashable/Class.hs
@@ -87,7 +87,7 @@ import qualified Data.ByteString.Lazy.Internal as BL  -- foldlChunks
 import qualified Data.ByteString.Short.Internal as BSI
 #endif
 
-#ifdef VERSION_integer_gmp
+#if defined(VERSION_integer_gmp) && !defined(ghcjs_HOST_OS)
 import GHC.Exts (Int(..))
 import GHC.Integer.GMP.Internals (Integer(..))
 # if MIN_VERSION_integer_gmp(1,0,0)
@@ -278,7 +278,7 @@ instance Hashable Natural where
 #endif
 
 instance Hashable Integer where
-#if defined(VERSION_integer_gmp)
+#if defined(VERSION_integer_gmp) && !defined(ghcjs_HOST_OS)
 # if MIN_VERSION_integer_gmp(1,0,0)
     hash (S# n)   = (I# n)
     hash (Jp# bn) = hash bn
