diff --git a/Data/Hashable/Class.hs b/Data/Hashable/Class.hs
index 1ab9879..b809004 100644
--- a/Data/Hashable/Class.hs
+++ b/Data/Hashable/Class.hs
@@ -87,7 +87,7 @@ import qualified Data.ByteString.Lazy.Internal as BL  -- foldlChunks
 import qualified Data.ByteString.Short.Internal as BSI
 #endif
 
-#ifdef VERSION_integer_gmp
+#if defined(VERSION_integer_gmp) && !defined(ghcjs_HOST_OS)
 import GHC.Exts (Int(..))
 import GHC.Integer.GMP.Internals (Integer(..))
 #endif
@@ -242,7 +242,7 @@ instance Hashable Char where
     hashWithSalt = defaultHashWithSalt
 
 instance Hashable Integer where
-#if defined(VERSION_integer_gmp)
+#if defined(VERSION_integer_gmp) && !defined(ghcjs_HOST_OS)
     hash (S# int) = I# int
     hash n@(J# size# byteArray)
         | n >= minInt && n <= maxInt = fromInteger n :: Int
