diff --git a/System/Process.hsc b/System/Process.hsc
index 37d442f..541497f 100644
--- a/System/Process.hsc
+++ b/System/Process.hsc
@@ -6,6 +6,10 @@
 #endif
 {-# LANGUAGE InterruptibleFFI #-}
 
+##ifdef ghcjs_HOST_OS
+{-# LANGUAGE JavaScriptFFI #-}
+##endif
+
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  System.Process
@@ -674,6 +678,22 @@ terminateProcess ph = do
 -- ----------------------------------------------------------------------------
 -- Interface to C bits
 
+##if defined(ghcjs_HOST_OS)
+
+foreign import javascript unsafe
+  "h$process_terminateProcess($1)"
+  c_terminateProcess :: PHANDLE -> IO CInt
+
+foreign import javascript unsafe
+  "h$process_getProcessExitCode($1,$2_1,$2_2)"
+  c_getProcessExitCode :: PHANDLE -> Ptr CInt -> IO CInt
+
+foreign import javascript interruptible
+  "h$process_waitForProcess($1,$2_1,$2_2,$c);"
+  c_waitForProcess :: PHANDLE -> Ptr CInt -> IO CInt
+
+##else
+
 foreign import ccall unsafe "terminateProcess"
   c_terminateProcess
         :: PHANDLE
@@ -691,6 +711,7 @@ foreign import ccall interruptible "waitForProcess" -- NB. safe - can block
         -> Ptr CInt
         -> IO CInt
 
+##endif
 
 -- ----------------------------------------------------------------------------
 -- Old deprecated variants
diff --git a/System/Process/Internals.hs b/System/Process/Internals.hs
index c4d5be3..dd59a82 100644
--- a/System/Process/Internals.hs
+++ b/System/Process/Internals.hs
@@ -3,6 +3,10 @@
 {-# LANGUAGE Trustworthy #-}
 {-# LANGUAGE InterruptibleFFI #-}
 
+#ifdef __GHCJS__
+{-# LANGUAGE ForeignFunctionInterface, JavaScriptFFI, UnliftedFFITypes, MagicHash #-}
+#endif
+
 -----------------------------------------------------------------------------
 -- |
 -- Module      :  System.Process.Internals
@@ -50,6 +54,11 @@ import System.Posix.Internals (FD)
 
 import System.Process.Common
 
+#ifdef ghcjs_HOST_OS
+import Control.Applicative
+import GHCJS.Prim
+#endif
+
 #if WINDOWS
 import System.Process.Windows
 #else
@@ -138,6 +147,46 @@ translate = translateInternal
 {-# INLINE translate #-}
 
 
+#if defined(ghcjs_HOST_OS)
+
+type JSArray  = JSVal
+type JSObject = JSVal
+type JSString = JSVal
+
+fromJSStrings :: JSVal -> IO [String]
+fromJSStrings x = fmap (map fromJSString) (fromJSArray x)
+
+fromJSInts :: JSVal -> IO [Int]
+fromJSInts x = map fromJSInt <$> fromJSArray x
+
+toJSStrings :: [String] -> IO JSVal
+toJSStrings xs = toJSArray (map toJSString xs)
+
+throwErrnoIfJSNull :: String -> IO JSVal -> IO JSVal
+throwErrnoIfJSNull msg m = do
+  r <- m
+  if isNull r then throwErrno msg
+              else return r
+
+foreign import javascript safe
+  "h$process_runInteractiveProcess($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)"
+  js_runInteractiveProcess :: JSString     -- ^ $1 command or program
+                           -> JSArray      -- ^ $2 arguments, null if it's a raw command
+                           -> JSString     -- ^ $3 working dir, null for current
+                           -> JSArray      -- ^ $4 environment, null for existing
+                           -> CInt         -- ^ $5 stdin fd
+                           -> CInt         -- ^ $6 stdout fd
+                           -> CInt         -- ^ $7 stderr fd
+                           -> Bool         -- ^ $8 close handles
+                           -> Bool         -- ^ $9
+                           -> Bool         -- ^ $10 delegate ctrl-c
+                           -> IO JSVal     -- ^ process handle
+
+foreign import javascript safe
+  "h$process_commandToProcess($1,$2)"
+  js_commandToProcess :: JSString -> JSArray -> IO JSArray
+#endif
+
 -- ----------------------------------------------------------------------------
 -- Deprecated / compat
 
diff --git a/process.cabal b/process.cabal
index 15f6a65..a944525 100644
--- a/process.cabal
+++ b/process.cabal
@@ -71,6 +71,9 @@ library
                    filepath  >= 1.2 && < 1.5,
                    deepseq   >= 1.1 && < 1.5
 
+    if(impl(ghcjs))
+      build-depends: ghcjs-prim
+
 test-suite test
   default-language: Haskell2010
   hs-source-dirs: test
